server/src/middleware/rpc/layer/logger.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! RPC Logger layer.
+
+use std::{
+	pin::Pin,
+	task::{Context, Poll},
+};
+
+use futures_util::Future;
+use jsonrpsee_core::{
+	server::MethodResponse,
+	tracing::{rx_log_from_json, tx_log_from_str},
+};
+use jsonrpsee_types::Request;
+use pin_project::pin_project;
+
+use crate::middleware::rpc::RpcServiceT;
+
+/// RPC logger layer.
+#[derive(Copy, Clone, Debug)]
+pub struct RpcLoggerLayer(u32);
+
+impl RpcLoggerLayer {
+	/// Create a new logging layer.
+	pub fn new(max: u32) -> Self {
+		Self(max)
+	}
+}
+
+impl<S> tower::Layer<S> for RpcLoggerLayer {
+	type Service = RpcLogger<S>;
+
+	fn layer(&self, service: S) -> Self::Service {
+		RpcLogger { service, max: self.0 }
+	}
+}
+
+/// A middleware that logs each RPC call and response.
+#[derive(Debug)]
+pub struct RpcLogger<S> {
+	max: u32,
+	service: S,
+}
+
+impl<'a, S> RpcServiceT<'a> for RpcLogger<S>
+where
+	S: RpcServiceT<'a> + Send + Sync + Clone + 'static,
```

Comment:
I geuss my comment on `Clone` should have been here and not on an example but whatever :)